skfp/model_selection/splitters/pubchem_split.py

```
@@ -0,0 +1,447 @@
+
+    try:
+        cid = response_json["IdentifierList"]["CID"][0]
+        cid = str(cid) if cid else None
+    except (KeyError, ValueError):
+        cid = None
+    return cid
+
+
+def _get_earliest_publication_date(
+    cid: Optional[str], n_trials: int = 5
+) -> Optional[int]:
+    """
+    Get the date of the earliest publication from PubChem where a given molecule appears.
+    There are molecules without publications, for which we return None.
+
+    """
+    if not cid:
+        return None
+
+    query = {
+        "download": "*",
+        "collection": "literature",
+        "where": {"ands": [{"cid": str(cid)}]},
+        "order": ["articlepubdate,asc"],
+        "start": 0,
+        "limit": 3,  # fetch a few, just in case
+        "downloadfilename": f"pubchem_cid_{cid}_literature",
+        "nullatbottom": 1,
+    }
+    base_url = "https://pubchem.ncbi.nlm.nih.gov/sdq/sdqagent.cgi"
+    params = {"infmt": "json", "outfmt": "json", "query": json.dumps(query)}
+
+    def get_publication_date(response: list[dict]) -> Optional[int]:
+        article_pub_date = None
+        for entry in response:
+            if "articlepubdate" in entry:
+                return int(entry["articlepubdate"][:4])  # get year of publication
+        return article_pub_date
+
+    trial = 0
+    while trial < n_trials:
+        try:
+            response = requests.get(base_url, params=params, timeout=10).json()
+            if isinstance(response, list):
+                if len(response) >= 1:
+                    date = get_publication_date(response)
+                    return date
+                else:
+                    # no literature found
+                    return None
+            else:
+                # most probably "Server too busy" error
+                time.sleep(1)
+        except requests.ConnectionError:
+            # most probably unstable PubChem network
+            time.sleep(1)
+        trial += 1
+
+    return None
```

If all trials failed, I think we should raise an error, because it's not `None`, but rather a failure because of PubChem servers